lib/datadog/appsec/utils/read_write_lock.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+module Datadog
+  module AppSec
+    module Utils
+      # Simple implentation for a fair read write lock
+      # It allows for as many concurrent readers but only one concurrent writer.
+      # To avoid the writer thread being exhausted, once a writer thread has signaled that it wants to acquire the lock,
+      # no more reader threads can acquire the lock.
+      # The new reader threads will wait for the writer thread to finish.
+      class ReadWriteLock
+        def initialize
+          @reader_mutex    = Mutex.new
+          @reader_q        = ConditionVariable.new
+          @reader_count    = 0
+          @reader_releases = 0
+
+          @writer_mutex    = Mutex.new
+          @writer_q        = ConditionVariable.new
+          @writter         = false
+        end
+
+        def with_rlock
+          rlock
+          yield
+        ensure
+          runlock
+        end
+
+        def with_lock
+          lock
+          yield
+        ensure
+          unlock
+        end
+
+        def rlock
+          @reader_mutex.synchronize do
+            @reader_count += 1
+
+            writter = false
+
+            @writer_mutex.synchronize do
+              writter = @writter
+            end
+
+            @reader_q.wait(@reader_mutex) if writter
+          end
+        end
+
+        def runlock
+          @reader_mutex.synchronize do
+            @reader_releases += 1
```

Comment:
The implementation never reset the reader and release count to 0. So answering your question, yes, it will go towards infinite. 

That could be an issue, as the value of reader count and release will keep growing and might lead to problems if the customer application is not deployed, releasing the values. 

We might be able to reset the reader and release count to 0 once the values are the same in the `rlock` method.

Reply:
Or as you suggested in the comment below, we could have a single variable `reader_count `